File: 089/gray-code.js
/**
 * @param {number} n
 * @return {number[]}
 */
var grayCode = function (n) {
    if (n <= 0) return [0];
    else {
        let prev = grayCode(n-1);
        Array.prototype.push.apply(prev, prev.map(i => i | (1 << (n-1))).reverse());
        return prev;
    }
};

File: 089/GrayCode.scala
object GrayCode {
  def grayCode(n: Int): List[Int] =
    if (n <= 0) List(0)
    else {
      val prev = grayCode(n - 1)
      prev ::: prev.map(i => i | (1<<(n-1)) ).reverse
    }

  def main(args: Array[String]): Unit = {
    var n = 2
    println(grayCode(n))
  }
}
